stdlib/SparseArrays/src/sparsematrix.jl

```diff
@@ -1423,23 +1371,15 @@ julia> sprand(Float64, 3, 0.75)
   [3]  =  0.298614
 ```
 """
-function sprand(r::AbstractRNG, m::Integer, n::Integer, density::AbstractFloat,
-                rfn::Function, ::Type{T}=eltype(rfn(r,1))) where T
-    m,n = Int(m), Int(n)
-    N = m*n
-    N == 0 && return spzeros(T,m,n)
-    N == 1 && return rand(r) <= density ? sparse([1], [1], rfn(r,1)) : spzeros(T,1,1)
-    _sprand(r,m,n,density,i->rfn(r,i))
+function sprand(r::AbstractRNG, m::Integer, n::Integer, density::AbstractFloat, rfn::Function, ::Type{T} = eltype(rfn(r, 1))) where T
+    m, n = Int(m), Int(n)
+    (m < 0 || n < 0) && throw(ArgumentError("invalid Array dimensions"))
+    0 <= density <= 1 || throw(ArgumentError("$density not in [0,1]"))
+    I = randsubseq(r, 1:(m*n), density)
+    sparse_sorted_unique!(I, rfn(r,length(I)), m, n)
```

Comment:
Shouldn't this be `T[rfn(r) for i=1:length(I)]`?

And then use `Type{T} = rfn(r)` above.  Otherwise `V` will have the wrong type if the caller passes a different `T` above. (We should have a test for this case.)

Reply:
Yes, this is a bug that is present (see #30627). My plan is to remove T entirely :smile: But you are right, we can at least fix this here.

Reply:
In the case where `T` wasn't passed, `rfn(r,length(I))` may be faster, since I think the `rand` function has optimizations for filling a whole array.

Reply:
I'll just use `convert(Vector{Tv},rfn(r,length(I))` since this is noop when it is already of the right type.

Reply:
if the user is worried about performance, he could wait for #30627 to be resolved, or use `(r,i)->rand(r,T,i)`. In any case we cannot solve this without breaking the interface. Either we leave the bug, or we make a copy.